AIS static data (class A and B) PGN support

adwuk opened the pull request.

ttlappalainen commented:
Hi,

Sorry to bother you again, but I prefer to define:
void SetN2kPGN129794(tN2kMsg &N2kMsg, uint8_t MessageID, tN2kAISRepeat Repeat, uint32_t UserID,  uint32_t IMOnumber, char *Callsign, char *Name, tN2kVeselType VesselType, double Length,  double Beam, double PosRefStbd, double PosRefBow, uint16_t ETAdate, double ETAtime,  double Draught, char *Destination, tN2kAISVersion AISversion, tN2kGNSStype GNSStype, bool DTE, tN2kAISInfo AISinfo)

tN2kGNSStype has been already defined, but other new types you have to write yourself. 
Also according my sources (pgn.h) N2kMsg.Add4ByteUInt(ETAtime); should be N2kMsg.Add4ByteUDouble(ETAtime,0.0001); so the ETAtime would be really seconds since midnight.

Check also other calls to use same new enum types and fix for ETAtime.

adwuk commented:
Hi - I will have a look at the enum types (I count 4 that you want added), but it won't be for a while, unless I can simply copy them from pgn.h.  I'll have a look.  I'll also check the ETAtime value.  It seemed to be working with the test data I had.  The project that I have been working on is done now, so unlikely that I make many more changes in the short term.

ttlappalainen commented:
You can use pgn.h as reference. Just change the types like in other enum types. So e.g. tN2kVesselType (note!I had typo error on name) will have value N2kvt_WingInGround=20. The vt on prefix N2kvt_ comes from "vessel type". With the same logic  tN2kAISRepeat values will have prefix N2kaisr_

ETAtime works if you feed uint_32 value 300000 for time 30 s after midnight. But we would like to use seconds since midnight, so you have to feed it in decimal 30.0000.

adwuk commented:
ETAtime I believe is now correct - don't know why it worked before, because it shouldn't have :)  Although it works in my test environment, I will give it another thorough test again at the weekend with live/real data.  I also changed the code to use the existing enum for GNSStype and added one for the AISRepeat variable.  There isn't a list for the AIS version in pgn.h so it will need to be figured out, and I am not convinced that the vessel type is exhaustive which may cause problems if a new type is identified.  There should also be an enum for NavStatus for example.  My suggestion is that all of these are added to a to-do list and dealt with at another stage.

ttlappalainen commented:
You are right about vessel type. It is long list to define as enum, byt the type should be uint8_t not int. Then we could have vessel types as defines:
#define N2kvt_WingInGround 20
etc.

I also noticed that DTE has been defined 0=Available and 1=Not Available. Meaning that your bool is opposite. So it should be:
N2kMsg.AddByte((DTE?0:64)  | (GNSStype &0x0f)<<2 | (AISversion & 0x03));

Accroding my data AIS version should be
enum tAISVersion {
                            N2kaisv_ITU_R_M_1371_1=0,
                            N2kaisv_ITU_R_M_1371_3=1,
                          };
enum tAISTranceiverInfo {
                            N2kaisti_Channel_A_VDL_reception=0,
                            N2kaisti_Channel_B_VDL_reception=1,
                            N2kaisti_Channel_A_VDL_transmission=2,
                            N2kaisti_Channel_B_VDL_transmission=3,
                            N2kaisti_Own_information_not_broadcast=4,
                            N2kaisti_Reserved=5
                          };

Where is NavStaus?

I'll prefer to make those changes before publish, because chaning interface is not nice.

adwuk commented:
additional enums now included and VesselType changed to uint8_t.  I disagree with changing the bool logic for DTE and others.  I defined them as bool to enforce a single bit value, not to make sense of the true/false values.  There are a few of them which don't really align with true/false, like the class B position report (DSC, Display, Msg22 etc.).  If we invert the logic as you suggest it means that the 1/0 values do not match the available documentation (i.e. http://catb.org/gpsd/AIVDM.html) and is not compatible with direct translation with NMEA 0183.

ttlappalainen commented:
I think that also with those it would be better to use enum instead of bool. In compiler bool anyway translates to byte depending of system. So by forcing then to enum, you also force possible values for setting. Also with enum all values makes sense and are more readable.

adwuk commented:
another enum added for DTE.

adwuk commented:
Done with enums now.  So feel free to merge these changes.
